// @strapi/core/dist/services/entity-validator/validators.js
"use strict";
const _ = require("lodash");
const strapiUtils = require("@strapi/utils");
const blocksValidator = require("./blocks-validator.js");
const _interopDefault = (e) => e && e.__esModule ? e : { default: e };
const ___default = /* @__PURE__ */ _interopDefault(_);
const addMinLengthValidator = (validator, {
  attr
}, { isDraft }) => {
  return attr.minLength && ___default.default.isInteger(attr.minLength) && !isDraft ? validator.min(attr.minLength) : validator;
};
const addMaxLengthValidator = (validator, {
  attr
}) => {
  return attr.maxLength && ___default.default.isInteger(attr.maxLength) ? validator.max(attr.maxLength) : validator;
};
const addMinIntegerValidator = (validator, {
  attr
}) => ___default.default.isNumber(attr.min) ? validator.min(___default.default.toInteger(attr.min)) : validator;
const addMaxIntegerValidator = (validator, {
  attr
}) => ___default.default.isNumber(attr.max) ? validator.max(___default.default.toInteger(attr.max)) : validator;
const addMinFloatValidator = (validator, {
  attr
}) => ___default.default.isNumber(attr.min) ? validator.min(attr.min) : validator;
const addMaxFloatValidator = (validator, {
  attr
}) => ___default.default.isNumber(attr.max) ? validator.max(attr.max) : validator;
const addStringRegexValidator = (validator, {
  attr
}) => {
  return "regex" in attr && !___default.default.isUndefined(attr.regex) ? validator.matches(new RegExp(attr.regex), { excludeEmptyString: !attr.required }) : validator;
};
const addUniqueValidator = (validator, {
  attr,
  model,
  updatedAttribute,
  entity,
  componentContext
}, options) => {
  if (attr.type !== "uid" && !attr.unique) {
    return validator;
  }
  return validator.test("unique", "This attribute must be unique", async (value) => {
    const isPublish = options.isDraft === false;
    if (___default.default.isNil(value)) {
      return true;
    }
    if (!isPublish && value === entity?.[updatedAttribute.name]) {
      return true;
    }
    let queryUid;
    let queryWhere = {};
    const hasPathToComponent = componentContext?.pathToComponent?.length > 0;
    if (hasPathToComponent) {
      const hasRepeatableData = componentContext.repeatableData.length > 0;
      if (hasRepeatableData) {
        const { name: updatedName, value: updatedValue } = updatedAttribute;
        const pathToCheck = [...componentContext.pathToComponent.slice(1), updatedName].join(".");
        const values = componentContext.repeatableData.map((item) => {
          return pathToCheck.split(".").reduce((acc, key) => acc[key], item);
        });
        const isUpdatedAttributeRepeatedInThisEntity = values.filter((value2) => value2 === updatedValue).length > 1;
        if (isUpdatedAttributeRepeatedInThisEntity) {
          return false;
        }
      }
      const {
        model: parentModel,
        options: parentOptions,
        id: excludeId
      } = componentContext.parentContent;
      queryUid = parentModel.uid;
      const whereConditions = {};
      const isParentDraft = parentOptions && parentOptions.isDraft;
      whereConditions.publishedAt = isParentDraft ? null : { $notNull: true };
      if (parentOptions?.locale) {
        whereConditions.locale = parentOptions.locale;
      }
      if (excludeId && !Number.isNaN(excludeId)) {
        whereConditions.id = { $ne: excludeId };
      }
      queryWhere = {
        ...componentContext.pathToComponent.reduceRight((acc, key) => ({ [key]: acc }), {
          [updatedAttribute.name]: value
        }),
        ...whereConditions
      };
    } else {
      queryUid = model.uid;
      const scalarAttributeWhere = {
        [updatedAttribute.name]: value
      };
      scalarAttributeWhere.publishedAt = options.isDraft ? null : { $notNull: true };
      if (options?.locale) {
        scalarAttributeWhere.locale = options.locale;
      }
      if (entity?.id) {
        scalarAttributeWhere.id = { $ne: entity.id };
      }
      queryWhere = scalarAttributeWhere;
    }
    return !await strapi.db.query(queryUid).findOne({ where: queryWhere });
  });
};
const stringValidator = (metas, options) => {
  let schema = strapiUtils.yup.string().transform((val, originalVal) => originalVal);
  schema = addMinLengthValidator(schema, metas, options);
  schema = addMaxLengthValidator(schema, metas);
  schema = addStringRegexValidator(schema, metas);
  schema = addUniqueValidator(schema, metas, options);
  return schema;
};
const emailValidator = (metas, options) => {
  const schema = stringValidator(metas, options);
  return schema.email().min(1, "${path} cannot be empty");
};
const uidValidator = (metas, options) => {
  const schema = stringValidator(metas, options);
  return schema.matches(/^[A-Za-z0-9-_.~]*$/);
};
const enumerationValidator = ({ attr }) => {
  return strapiUtils.yup.string().oneOf((Array.isArray(attr.enum) ? attr.enum : [attr.enum]).concat(null));
};
const integerValidator = (metas, options) => {
  let schema = strapiUtils.yup.number().integer();
  schema = addMinIntegerValidator(schema, metas);
  schema = addMaxIntegerValidator(schema, metas);
  schema = addUniqueValidator(schema, metas, options);
  return schema;
};
const floatValidator = (metas, options) => {
  let schema = strapiUtils.yup.number();
  schema = addMinFloatValidator(schema, metas);
  schema = addMaxFloatValidator(schema, metas);
  schema = addUniqueValidator(schema, metas, options);
  return schema;
};
const bigintegerValidator = (metas, options) => {
  const schema = strapiUtils.yup.mixed();
  return addUniqueValidator(schema, metas, options);
};
const datesValidator = (metas, options) => {
  const schema = strapiUtils.yup.mixed();
  return addUniqueValidator(schema, metas, options);
};
const validators = {
  string: stringValidator,
  text: stringValidator,
  richtext: stringValidator,
  password: stringValidator,
  email: emailValidator,
  enumeration: enumerationValidator,
  boolean: () => strapiUtils.yup.boolean(),
  uid: uidValidator,
  json: () => strapiUtils.yup.mixed(),
  integer: integerValidator,
  biginteger: bigintegerValidator,
  float: floatValidator,
  decimal: floatValidator,
  date: datesValidator,
  time: datesValidator,
  datetime: datesValidator,
  timestamp: datesValidator,
  blocks: blocksValidator
};
module.exports = validators;
